From dec62cd920161a792dcd563b8955c4ecd771d041 Mon Sep 17 00:00:00 2001
From: Francesco VIRLINZI <francesco.virlinzi@st.com>
Date: Wed, 6 May 2009 12:20:31 +0200
Subject: [PATCH] stm: pm: Registered EMI as sysdev device

This patch registers the EMI as sysdev device to better support
the power management

Signed-off-by: Francesco Virlinzi <francesco.virlinzi@st.com>
---
 drivers/stm/emi.c |   31 ++++++++++++++++++++++++++++++-
 1 files changed, 30 insertions(+), 1 deletions(-)

diff --git a/drivers/stm/emi.c b/drivers/stm/emi.c
index 6b25737..4ea15d1 100644
--- a/drivers/stm/emi.c
+++ b/drivers/stm/emi.c
@@ -7,6 +7,7 @@
  */
 
 #include <linux/kernel.h>
+#include <linux/sysdev.h>
 #include <linux/device.h>
 #include <linux/delay.h>
 #include <linux/io.h>
@@ -249,7 +250,7 @@ struct emi_pm {
 	struct emi_pm_bank bank[emi_num_bank];
 };
 
-int emi_pm_state(pm_message_t state)
+static int emi_sysdev_suspend(struct sys_device *dev, pm_message_t state)
 {
 	int idx;
 	int bank, data;
@@ -310,4 +311,32 @@ int emi_pm_state(pm_message_t state)
 	}
 	return 0;
 }
+
+static int emi_sysdev_resume(struct sys_device *dev)
+{
+	return emi_sysdev_suspend(dev, PMSG_ON);
+}
+
+static struct sysdev_class emi_sysdev_class = {
+	set_kset_name("emi"),
+};
+
+static struct sysdev_driver emi_sysdev_driver = {
+	.suspend = emi_sysdev_suspend,
+	.resume = emi_sysdev_resume,
+};
+
+struct sys_device emi_sysdev_dev = {
+	.cls = &emi_sysdev_class,
+};
+
+static int __init emi_sysdev_init(void)
+{
+	sysdev_class_register(&emi_sysdev_class);
+	sysdev_driver_register(&emi_sysdev_class, &emi_sysdev_driver);
+	sysdev_register(&emi_sysdev_dev);
+	return 0;
+}
+
+module_init(emi_sysdev_init);
 #endif
-- 
1.6.0.6

